staging/src/k8s.io/client-go/tools/events/event_recorder.go

```diff
@@ -67,7 +67,7 @@ func (recorder *recorderImpl) makeEvent(refRegarding *v1.ObjectReference, refRel
 	t := metav1.Time{Time: recorder.clock.Now()}
 	namespace := refRegarding.Namespace
 	if namespace == "" {
-		namespace = metav1.NamespaceDefault
+		namespace = metav1.NamespaceSystem
```

Comment:
Actually - we're using NamespaceDefault in the old events library:
https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/client-go/tools/record/event.go#L365-L368

The question is what we want do to here:
1) Change to NamespaceSystem here (as you do)
2) Change the validation logic to actually validate NamespaceDefault:
https://github.com/kubernetes/kubernetes/blob/master/pkg/apis/core/validation/events.go#L143

(1) might be misleading to users, as events that were previously landing in default namespace, will now be landing system namespace.
OTOH (2) seems more natural to me (I think that kube-system is more namespace for Node-related events). It may not be the case for other cluster-scoped resources though (e.g. PVs).

Neither of those options is perfect, but my personal preference would be towards (1) as we should prefer consistency (if someone is used to have events in default namespace for last 20 releases, we probably shouldn't change that now).

But I would like to hear @liggitt opinion from API perspective (can we change validation logic)?

Reply:
I think `events.k8s.io/v1` events for cluster-scoped resources should be created in the same namespace that core `v1` events were.

We should not tighten API validation in a way that would break existing clients that are successfully writing events.

First, add a standalone integration test that does the following:

1. exercises the following test matrix for an event for a cluster-scoped object:

  * operations: `create`, `update`
  * API versions: `v1`, `events.k8s.io/v1`
  * namespaces: `default`, `kube-system`

2. uses the core v1 and events.k8s.io event recorders to record an event for a cluster-scoped object

Then, update the events.k8s.io client to behave like the core v1 client, creating events in the same namespace.

Reply:
Thanks Jordan.
I now realized that I wrote:

>  but my personal preference would be towards (1) as we should prefer consistency (if someone is used to have events in default namespace for last 20 releases, we probably shouldn't change that now).

which doesn't make sense - I meant option (2) [the argument about consistency remains - it was for option (2)].

So basically - we don't need any client changes - the client libraries are consistent with each other.

We need to update the validation logic, because currently we enforce "kube-system", whereas both the old core/v1.Event and events.k8s.io/v1/Event are writing events to "default" namespace.

Reply:
agreed. we still need the tests to prove we're not breaking requests that work today, and that our clients actually work.

Reply:
Sure - we absolutely need tests. I just wanted to ensure that we will be testing the right thing - the current change isn't what we want.